@here/harp-mapview/lib/image/MipMapGenerator.ts

```diff
@@ -0,0 +1,159 @@
+        if (isNode) {
+            throw new Error("MipMapGenerator only works in browser.");
+        }
+
+        if (image.imageData === undefined) {
+            throw new Error("Can not generate mip maps. Image data not loaded!");
+        }
+        const imageData = image.imageData;
+        const mipMaps: ImageData[] = [];
+
+        // Add initial texture with padding as level 0
+        const { width: paddedWidth, height: paddedHeight } = MipMapGenerator.getPaddedSize(
+            imageData.width,
+            imageData.height
+        );
+        this.copyImageWithPadding(imageData, paddedWidth, paddedHeight);
+        mipMaps.push(this.m_paddingContext!.getImageData(0, 0, paddedWidth, paddedHeight));
+
+        let width = paddedWidth * 0.5;
+        let height = paddedHeight * 0.5;
+        while (width >= 1 && height >= 1) {
+            const mipMapLevel = mipMaps.length;
+            const previousImage = mipMaps[mipMapLevel - 1];
+            // Resize previous mip map level
+            mipMaps.push(this.resizeImage(previousImage, width, height));
+            width *= 0.5;
+            height *= 0.5;
+        }
+
+        return mipMaps;
+    }
+
+    /**
+     * Copy image to a canvas and add padding if necessary.
+     * @param image Input image.
+     * @param width Width of output image
+     * @param height Width of output image
+     * @returns Canvas with image and padding.
+     */
+    private copyImageWithPadding(
+        image: ImageData | ImageBitmap,
+        width: number,
+        height: number
+    ): HTMLCanvasElement {
+        this.m_paddingCanvas!.width = width;
+        this.m_paddingCanvas!.height = height;
+
+        this.m_paddingContext!.clearRect(0, 0, width, height);
+        if (image instanceof ImageBitmap) {
+            this.m_paddingContext!.drawImage(image, 0, 0);
+        } else {
+            this.m_paddingContext!.putImageData(image, 0, 0);
+        }
+
+        // Add horizontal padding
+        if (image.width !== width) {
+            this.m_paddingContext!.drawImage(
+                this.m_paddingCanvas!,
+                image.width - 1,
+                0,
+                1,
+                image.height,
+                image.width,
+                0,
+                width - image.width,
+                image.height
+            );
+        }
+
+        // Add vertical padding
+        if (image.height !== height) {
+            this.m_paddingContext!.drawImage(
+                this.m_paddingCanvas!,
+                0,
+                image.height - 1,
+                width,
+                1,
+                0,
+                image.height,
+                width,
+                height - image.height
+            );
+        }
+
+        return this.m_paddingCanvas!;
+    }
+
+    /**
+     * Resize an image. Quality of resized image is best when
+     * width = image.width / 2 and height = image.height / 2.
```

Comment:
I guess you mean quality is best when it is image.width % 2 === 0 ?

Reply:
Yes, that and also that an image should be resized exactly by factor 0.5 or 2. Those are the only use cases where the canvas API uses "correct" down/upsampling on most browsers. Otherwise it uses a faster approximation that has a worse image quality.

Reply:
Yes!